gnovm/stdlibs/frame.go

```diff
@@ -18,3 +21,39 @@ func (r Realm) PkgPath() string {
 func (r Realm) IsUser() bool {
 	return r.pkgPath == ""
 }
+
+// isOriginCall returns true if the std.OrigCaller == std.PrevRealm
+func isOriginCall(m *gno.Machine) bool {
+	return prevRealm(m).addr == m.Context.(ExecContext).OrigCaller
+}
+
+// prevRealm loops on frames and returns the second realm found in the calling
+// order. If no such realm was found, returns the tx signer (aka OrigCaller).
+func prevRealm(m *gno.Machine) Realm {
+	var lastRealmPath string
+	for i := m.NumFrames() - 1; i > 0; i-- {
+		fr := m.Frames[i]
+		if fr.LastPackage == nil || !fr.LastPackage.IsRealm() {
+			// Ignore non-realm frame
+			continue
+		}
+		realmPath := fr.LastPackage.PkgPath
+		if lastRealmPath == "" {
+			// Record the path of the first ecountered realm and continue
+			lastRealmPath = realmPath
+			continue
+		}
+		if lastRealmPath != realmPath {
+			// Second realm detected, return it.
+			return Realm{
+				addr:    fr.LastPackage.GetPkgAddr().Bech32(),
+				pkgPath: realmPath,
+			}
+		}
+	}
+	// No second realm found, return the tx signer.
+	return Realm{
+		addr:    m.Context.(ExecContext).OrigCaller,
+		pkgPath: "", // empty for users
+	}
+}
```

Comment:
Having a standalone function also improves the readability of the algorithm, which used to be quite confusing IMO :)